pkg/controller/podautoscaler/replica_calculator.go

```diff
@@ -252,26 +252,26 @@ func (c *ReplicaCalculator) GetObjectMetricReplicas(currentReplicas int32, targe
 		return 0, 0, time.Time{}, fmt.Errorf("unable to calculate ready pods: %s", err)
 	}
 
-	replicaCount = int32(math.Ceil(usageRatio * float64(readyPodCount)))
+	replicaCount = newSpecReplicas(int32(math.Ceil(usageRatio * float64(readyPodCount))))
 
 	return replicaCount, utilization, timestamp, nil
 }
 
 // GetObjectPerPodMetricReplicas calculates the desired replica count based on a target metric utilization (as a milli-value)
 // for the given object in the given namespace, and the current replica count.
-func (c *ReplicaCalculator) GetObjectPerPodMetricReplicas(currentReplicas int32, targetAverageUtilization int64, metricName string, namespace string, objectRef *autoscaling.CrossVersionObjectReference, metricSelector labels.Selector) (replicaCount int32, utilization int64, timestamp time.Time, err error) {
+func (c *ReplicaCalculator) GetObjectPerPodMetricReplicas(specReplicas specReplicas, targetAverageUtilization int64, metricName string, namespace string, objectRef *autoscaling.CrossVersionObjectReference, metricSelector labels.Selector) (replicaCount specReplicas, utilization int64, timestamp time.Time, err error) {
 	utilization, timestamp, err = c.metricsClient.GetObjectMetric(metricName, namespace, objectRef, metricSelector)
 	if err != nil {
 		return 0, 0, time.Time{}, fmt.Errorf("unable to get metric %s: %v on %s %s/%s", metricName, objectRef.Kind, namespace, objectRef.Name, err)
 	}
 
-	replicaCount = currentReplicas
+	replicaCount = specReplicas
 	usageRatio := float64(utilization) / (float64(targetAverageUtilization) * float64(replicaCount))
 	if math.Abs(1.0-usageRatio) > c.tolerance {
 		// update number of replicas if change is large enough
-		replicaCount = int32(math.Ceil(float64(utilization) / float64(targetAverageUtilization)))
+		replicaCount = newSpecReplicas(int32(math.Ceil(float64(utilization) / float64(targetAverageUtilization))))
 	}
-	utilization = int64(math.Ceil(float64(utilization) / float64(currentReplicas)))
+	utilization = int64(math.Ceil(float64(utilization) / float64(specReplicas)))
```

Comment:
I'm not sure that this is a good idea.
I mean, scale.Status.Replicas and scale.Spec.Replicas might change a lot.
If I'm not mistaken these numbers might differ a lot. And in this case the utilization here ^^ will be calculated incorrectly.

Reply:
You are correct.  I'm using the wrong denominator.  I'll plumb statusReplicas back through.

Reply:
And my use of `replicaCount` (which is `specReplicas`) when calculating the original `usageRatio` is wrong too.  I don't think `specReplicas` should ever be in the denominator of a ratio calculation because it doesn't relate directly to the metrics and pods found.  Does that sound right?

Reply:
@gliush I've been thinking about the per-pod object metrics use case.  And trying to make the unit tests make sense.

If I divide by `statusReplicas` then the utilization changes enough to exceed the tolerance and enter the conditional to update `replicaCount`.  However the calculation divides the object metric by the configured target, not taking replica counts into account.  So it will always come up with the `statusReplicas` (and maybe +1 because of the ceiling).  There is nothing else for it to do because adding or removing Pods will not change the object metric.

So dividing by `statusReplicas` is pointless.  In this particular case, I think dividing by `statusReplicas` is the right thing to do.

Reply:
>  I don't think specReplicas should ever be in the denominator of a ratio calculation because it doesn't relate directly to the metrics and pods found. 

Right.

> However the calculation divides the object metric by the configured target, not taking replica counts into account. 

It does take it into account. Let me show it:
1. `usageRatio := float64(utilization) / float64(targetUtilization)`
https://github.com/kubernetes/kubernetes/blob/b8b1720f127f9a4e6784bc652faafb256322353f/pkg/controller/podautoscaler/replica_calculator.go#L243

2. `replicaCount = int32(math.Ceil(usageRatio * float64(readyPodCount)))`
https://github.com/kubernetes/kubernetes/blob/b8b1720f127f9a4e6784bc652faafb256322353f/pkg/controller/podautoscaler/replica_calculator.go#L255

3. `return replicaCount ...`

I.e. the `utilizationRatio` is indeed calculated without the current number of pods. But the desired number of pods is calculated by multiplying the current number of pods by that `utilizationRatio`. That is where the current `statusReplicas` is used.

Another argument, more conceptual.
The idea of HPA itself is to autoscale pods under control of some metric value.
It means that we should always consider current load to be influenced (formed) by a current number of pods. And that increased number of pods will reduce the metric value.

As a result, it doesn't make sense to use the `specReplicas` in any calculation, because current state depends only on `statusReplicas`.

Reply:
Your example above is a link to `GetObjectMetricReplicas` which does indeed consider pod count.  So of course we should divide by `statusReplicas` or `readyPodCount`.

https://github.com/kubernetes/kubernetes/blob/b8b1720f127f9a4e6784bc652faafb256322353f/pkg/controller/podautoscaler/replica_calculator.go#L255

However your initial comment and the interesting case here is in `GetObjectPerPodMetricReplicas`.  We don't divide by the number of pods.  Just the metric by the target.

https://github.com/kubernetes/kubernetes/blob/b8b1720f127f9a4e6784bc652faafb256322353f/pkg/controller/podautoscaler/replica_calculator.go#L272